### ISSUE TITLE
NestedLoopJoin will panic when right child contains RepartitionExec

### ISSUE THREAD

**Issue by ygf11** [apache/arrow-datafusion#5022]
**Describe the bug**
When I work on #4866, I find some `NestedLoopJoin`s will panic.
```sql
# create tables -- t.csv is an empty file.
❯ CREATE EXTERNAL TABLE t1 (t1_id INT, t1_name text, t1_int INT) STORED AS CSV LOCATION 't.csv';
❯ CREATE EXTERNAL TABLE t2 (t2_id INT, t2_name text, t2_int INT) STORED AS CSV LOCATION 't.csv';
# target_partitions
> set datafusion.execution.target_partitions = 4;
```

```sql
❯ explain select * from t1 inner join t2 on t1.t1_id > t2.t2_id where t1.t1_id > 10 and t2.t2_int > 10;
+---------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                                                                  |
+---------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: t1.t1_id, t1.t1_name, t1.t1_int, t2.t2_id, t2.t2_name, t2.t2_int                                                                                          |
|               |   Inner Join:  Filter: t1.t1_id > t2.t2_id                                                                                                                            |
|               |     Filter: t1.t1_id > Int32(10)                                                                                                                                      |
|               |       TableScan: t1 projection=[t1_id, t1_name, t1_int], partial_filters=[t1.t1_id > Int32(10)]                                                                       |
|               |     Filter: t2.t2_int > Int32(10)                                                                                                                                     |
|               |       TableScan: t2 projection=[t2_id, t2_name, t2_int], partial_filters=[t2.t2_int > Int32(10)]                                                                      |
| physical_plan | ProjectionExec: expr=[t1_id@0 as t1_id, t1_name@1 as t1_name, t1_int@2 as t1_int, t2_id@3 as t2_id, t2_name@4 as t2_name, t2_int@5 as t2_int]                         |
|               |   NestedLoopJoinExec: join_type=Inner, filter=BinaryExpr { left: Column { name: "t1_id", index: 0 }, op: Gt, right: Column { name: "t2_id", index: 1 } }              |
|               |     CoalesceBatchesExec: target_batch_size=8192                                                                                                                       |
|               |       FilterExec: t1_id@0 > 10                                                                                                                                        |
|               |         RepartitionExec: partitioning=RoundRobinBatch(4), input_partitions=1                                                                                          |
|               |           CsvExec: files={1 group: [[home/work/tools/datafusion-test-data/join-context/t1.csv]]}, has_header=false, limit=None, projection=[t1_id, t1_name, t1_int]   |
|               |     CoalescePartitionsExec                                                                                                                                            |
|               |       CoalesceBatchesExec: target_batch_size=8192                                                                                                                     |
|               |         FilterExec: t2_int@2 > 10                                                                                                                                     |
|               |           RepartitionExec: partitioning=RoundRobinBatch(4), input_partitions=1 -- panic in RepartitionExec                                                                                          |
|               |             CsvExec: files={1 group: [[home/work/tools/datafusion-test-data/join-context/t1.csv]]}, has_header=false, limit=None, projection=[t2_id, t2_name, t2_int] |
|               |                                                                                                                                                                       |
+---------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

This sql will panic:
```sql
❯ select * from t1 inner join t2 on t1.t1_id > t2.t2_id where t1.t1_id > 10 and t2.t2_int > 10;
thread 'tokio-runtime-worker' panicked at 'partition not used yet', /home/work/arrow-datafusion/datafusion/core/src/physical_plan/repartition/mod.rs:426:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'tokio-runtime-worker' panicked at 'partition not used yet', /home/work/arrow-datafusion/datafusion/core/src/physical_plan/repartition/mod.rs:426:14
thread 'tokio-runtime-worker' panicked at 'partition not used yet', /home/work/arrow-datafusion/datafusion/core/src/physical_plan/repartition/mod.rs:426:14
thread 'tokio-runtime-worker' panicked at 'partition not used yet', /home/work/arrow-datafusion/datafusion/core/src/physical_plan/repartition/mod.rs:426:14
thread 'tokio-runtime-worker' panicked at 'partition not used yet', /home/work/arrow-datafusion/datafusion/core/src/physical_plan/repartition/mod.rs:426:14
thread 'tokio-runtime-worker' panicked at 'partition not used yet', /home/work/arrow-datafusion/datafusion/core/src/physical_plan/repartition/mod.rs:426:14
thread 'tokio-runtime-worker' panicked at 'partition not used yet', /home/work/arrow-datafusion/datafusion/core/src/physical_plan/repartition/mod.rs:426:14
thread 'tokio-runtime-worker' panicked at 'partition not used yet', /home/work/arrow-datafusion/datafusion/core/src/physical_plan/repartition/mod.rs:426:14
thread 'tokio-runtime-worker' panicked at 'partition not used yet', /home/work/arrow-datafusion/datafusion/core/src/physical_plan/repartition/mod.rs:426:14
0 rows in set. Query took 0.015 seconds.
```

**To Reproduce**
See above.

**Expected behavior**
The join should work.

It seems the partition of right child may execute one more time in `NestedLoopJoin`, if it has multiple output partitions. It is ok in most time, but when meets `RepartitionExec` it will panic. 

  
**Additional context**

**Comment by liukun4515** [comment]
@ygf11 Do you find the root cause why the panic happened?

```
It seems the partition of right child may execute one more time in NestedLoopJoin, if it has multiple output partitions. It is ok in most time, but when meets RepartitionExec it will panic.
```

From your example, the left side is partitioned and the right side is required single-partitioned.

The output partition should be consistent with left side.

The right single-partition will be executed one more time, is it not allowed to be called multiple times?  Maybe it is the point why it caused panic
@alamb

**Comment by ygf11** [comment]
> The right single-partition will be executed one more time,

@liukun4515, that is the cause.

**Comment by liukun4515** [comment]
> ```sql
> RepartitionExec
> ```

Yes, we got the same reason, but I don't know why the `RepartitionExec` can not be called twice for the same partition.
Other physicalexec have the same behavior? @alamb

**Comment by liukun4515** [comment]
> > ```sql
> > RepartitionExec
> > ```
> 
> Yes, we got the same reason, but I don't know why the `RepartitionExec` can not be called twice for the same partition. Other physicalexec have the same behavior?

**Comment by liukun4515** [comment]
The `Repartition Exec` share a global state of 
```    
/// Inner state that is initialized when the first output stream is created.
    state: Arc<Mutex<RepartitionExecState>>,
```

the state will shared between multiple runtime/coroutine context, if one coroutine execute the `partition 1`, and other coroutine can not execute the `partition 1` again.

**Comment by ygf11** [comment]
> Other physicalexec have the same behavior?

I don't know if there is other physical plan has the same behavior.
but maybe executing the same partition one more time is not a good idea, it will waste a lot of time, a better way is to cache the result(`OnceFut`), or only call once.

`NestedLoopJoinExec` will sometimes build-left, and sometimes build-right now, currently they are using the same logic, the right way is  to give `build-left` a logic to handle join, and give `build-right` another logic.

The bug of example is the case of `build-right`, I see the `NestedLoopJoinStream` is using `build-left` logic.

I think maybe the it is an another cause. @liukun4515 @alamb

**Comment by ygf11** [comment]
> NestedLoopJoinExec will sometimes build-left, and sometimes build-right now, and they are using the same logic, I think a better way is to give build-left a logic to handle join, and give build-right another logic. Then the partitions of the non-build-side can be executed only once.

Let me explain more.

The main difference is the way of iteration.
For `build-left`, left has a single partition, right has multiple partitions. The iteration will be like:
```rust
// Given a partition number - x 
// execute right-partition-x
for batch in right-partition-x {
   join(batch, left-data-(single partition)) 
}
```

For `build-right`, left has multiple partitions, right has a single partition. The iteration will be like:
```rust
// Given a partition number - x 
// execute left-partition-x
for batch in left-partition-x {
   join(batch, right-data(single partition)) 
}
```

If we use `build-left` logic for `build-right`(issue-example), then:
```rust
// Given a partition number - x 
// Execute right-partition-x --- right partition is a single partition and execute many times
for batch in right-partition-x {
   join(batch, left-data) 
}
```

**Comment by liukun4515** [comment]
> > Other physicalexec have the same behavior?
> 
> I don't know if there is other physical plan has the same behavior. but maybe executing the same partition one more time is not a good idea, it will waste a lot of time, a better way is to cache the result(`OnceFut`), or only call once.
> 
> `NestedLoopJoinExec` will sometimes build-left, and sometimes build-right now, and they are using the same logic, I think a better way is to give `build-left` a logic to handle join, and give `build-right` another logic. Then the partitions of the `non-build-side` can be executed only once.
> 
> The bug of example is the case of `build-right`, I see the `NestedLoopJoinStream` is using `build-left` logic.
> 
> I think maybe it is an another cause. @liukun4515 @alamb 

I think the `build-side` and `probe-side` are the concept for the Hash-Join.

In the NLJ, we need the outer-side/outer table and inner-side/inner table.

```
  for out-row in outer-table
    for inner-row in inner-table
       check-join
```

The inner table will be traveled many times in the most basic implementation.

In the current implementation, the left table is the outer table, the right table is the inner table.

For the `inner join` with multi left partition and single right partition:

```
for left-partition-x in left multi partition:
   join(load-left(left-partition-x), load-right(single-right-partition))
```

I think this implementation of iter is matched with the algorithm of the NLJ.

**Comment by liukun4515** [comment]
I think we can change the function `distribution_from_join_type`
```
        JoinType::Inner | JoinType::Left | JoinType::LeftSemi | JoinType::LeftAnti => {
            // need the left data, and the right should be one partition
            vec![
                Distribution::UnspecifiedDistribution,
                Distribution::SinglePartition,
            ]
        }
```

to 

```
JoinType::Inner | JoinType::Left | JoinType::LeftSemi | JoinType::LeftAnti => {
            // need the left data, and the right should be one partition
            vec![
                Distribution::SinglePartition,
                Distribution::SinglePartition,
            ]
        }
```
temporarily If we can not support executing multi times for one physical executor for the same partition.

It will not block your fix for the issue https://github.com/apache/arrow-datafusion/pull/4866

cc @ygf11

**Comment by ygf11** [comment]
Thanks for response @liukun4515. I think I find the way to fix this bug.

> In the current implementation, the left table is the outer table, the right table is the inner table.

IMOP, The relationship between `inner/outer table` and `left/right table` is not fixed, it is decided by the required distribution(join type).
For `right/right-semi/right-anti` joins, 
* left is the single partition side.
* right is the multiple partition side.
* the output partition count is the same as right table.

The algorithm will be like:
```
for x in 0..output-partition-count:
   join(right-partition(x), single-left-partition)
```
we can see left table will be visited one more time, and right table will be visited only once.
That means right is the outer table side, and left is the inner table side in nested loop join.

> The inner table will be traveled many times in the most basic implementation.

Yes, visiting a relation many time is a common operation. 
In datafusion, other physical plans also have similar requirement, like `CrossJoinExec`, it works well with `OnceFut`. 

Currently the implementation will always add `OnceFut` for the left table, which is not correct(left table maybe outer table side).
I think if we correct add `OnceFut` for the inner table, the queries will success.

I create a new pr #5156 to fix this.
